File: Algorithm/10/findContentChildren.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int findContentChildren(vector<int>& g, vector<int>& s) {
        sort(g.begin(), g.end(), greater<int>());
        sort(s.begin(), s.end(), greater<int>());

        int si = 0, gi = 0;
        int res = 0;
        while(gi < g.size() && si < s.size()){
            if(s[si] >= g[gi]){
                res++;
                si++;
                gi++;
            }
            else
                gi++;
        }
        return res;
    }
};

class Solution1 {
public:
    int findContentChildren(vector<int>& g, vector<int>& s) {
        sort(g.begin(), g.end());
        sort(s.begin(), s.end());

        int si = s.size()-1, gi = g.size()-1;
        int res = 0;
        while(gi >= 0 && si >= 0){
            if(s[si] >= g[gi]){
                res++;
                si--;
                gi--;
            }
            else
                gi--;
        }
        return res;
    }
};

int main(){
    int g1[] = {1, 2, 3};
    vector<int> g(g1, g1 + sizeof(g1)/sizeof(int));
    int s1[] = {1, 1};
    vector<int> s(s1, s1 + sizeof(s1)/sizeof(int));
    int res = Solution1().findContentChildren(g,s);
    cout << res << endl;
    return 0;
}

File: Algorithm/10/findContentChildren.py
class Solution(object):
    def findContentChildren(self, g, s):
        """
        :type g: List[int]
        :type s: List[int]
        :rtype: int
        """
        res = 0
        g = sorted(g, reverse=True)
        s = sorted(s, reverse=True)
        while g and s:
            maxs = s[0]
            maxg = g[0]
            if maxs >= maxg:
                res += 1
                s.pop(0)
                g.pop(0)
            else:
                g.pop(0)
        return res

class Solution1(object):
    def findContentChildren(self, g, s):
        """
        :type g: List[int]
        :type s: List[int]
        :rtype: int
        """
        res = 0
        g = sorted(g)
        s = sorted(s)
        while g and s:
            maxs = s[-1]
            maxg = g[-1]
            if maxs >= maxg:
                res += 1
                s.pop()
                g.pop()
            else:
                g.pop()
        return res

class Solution2(object):
    def findContentChildren(self, g, s):
        """
        :type g: List[int]
        :type s: List[int]
        :rtype: int
        """
        res = 0
        g = sorted(g, reverse=True)
        s = sorted(s, reverse=True)
        si, gi = 0, 0
        while gi < len(g) and si < len(s):
            if  s[si] >= g[gi]:
                res += 1
                si += 1
                gi += 1
            else:
                gi += 1
        return res

if __name__ == '__main__':
    g = [1,2,3]
    s = [1,1]
    print(Solution2().findContentChildren(g,s))
